File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        vector<int>mpp (256, -1);
        int l = 0 , r = 0;
        int maxi = 0, n = s.size();

        while(r < n){
            if(mpp[s[r]] != -1){
                l = max(mpp[s[r]] + 1, l);
            }
            mpp[s[r]] = r;
            maxi = max(maxi , r - l + 1);
            r++;
        }
        return maxi;
    }
};

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        chars = set()
        left = 0
        maxi = 0

        for right in range(len(s)):
            while s[right] in chars:
                chars.remove(s[left])
                left += 1

            chars.add(s[right])
            maxi = max(maxi, right - left + 1)

        return maxi
